### PR TITLE
Make ephemerons immutable.

### PR DESCRIPTION
**Pull request by kayceesrk**
This is a port of https://github.com/ocaml/ocaml/pull/10737 for multicore. Compared to https://github.com/ocaml/ocaml/pull/10737 it removes all the deprecated functions.

### CODE REVIEW

**Comment by kayceesrk** [comment]
The consensus is that we keep the `Bucket` module in.

**Comment by kayceesrk** [comment]
@ctk21 this PR should be good to go except for a quick review.

**Review by ctk21 — COMMENTED** [review]
I took a (time limited) look. The PR does not alter the underlying runtime ephemeron GC mechanisms. However I think it does remove deprecated functions and makes the OCaml interface match up with post ocaml/ocaml#10737

I spotted a minor nit: the PR still references to the removed `iter`, `fold` functions in the documentation strings in `ephemeron.ml`.

I did not deeply review the testsuite alterations.

**Comment by ctk21** [comment]
I have further taken a look at the alterations to `ephetest_par.ml`. I believe the alterations are the correct ones in this case; that is I didn't spot anything where we have changed the behaviour of the test with this PR.

**Comment by kayceesrk** [comment]
Thanks @ctk21. Removed mentions of `iter` and `fold*` from documentation strings.

**Comment by kayceesrk** [comment]
Let me know if anything else looks odd..

**Comment by ctk21** [comment]
LGTM, merging!

File: testsuite/tests/weak-ephe-final/ephetest_par.ml [hunk 1]
```diff
@@ -9,6 +9,9 @@ open Ephemeron
 
 let debug = false
 
+open Printf
+open Ephemeron
+
```
Comment (damiendoligez) [inline testsuite/tests/weak-ephe-final/ephetest_par.ml:14]: Redundant opens.
```suggestion
```

**Review by damiendoligez — APPROVED** [review]
Looks good except for a small hiccup in `ephetest_par.ml`.

File: testsuite/tests/weak-ephe-final/ephetest_par.ml (hunk 1 above)
Comment (kayceesrk) [reply testsuite/tests/weak-ephe-final/ephetest_par.ml:14]: Fixed by #804.

**Pull request merged**